src/destructors.md

```diff
@@ -345,10 +345,14 @@ let x = (&temp()).use_temp();  // ERROR
 
 ## Not running destructors
 
-Not running destructors in Rust is safe even if it has a type that isn't
-`'static`. [`std::mem::ManuallyDrop`] provides a wrapper to prevent a
+[`std::mem::forget`] can be used to prevent the destructor of a variable from being run,
+and [`std::mem::ManuallyDrop`] provides a wrapper to prevent a
 variable or field from being dropped automatically.
 
+> Note: Preventing a destructor from being run via `forget` or other means is safe in Rust
+> even if it has a type that isn't `'static`. This means that publicly exposed APIs cannot
+> rely on destructor being run for soundness.
```

Comment:
It seems like it might be good to expand a little on this: for one thing, publicly exposed APIs seems like a somewhat "interesting" term. Certainly even in a crate-local context I would expect to see unsafe fn on the constructor or similar if the destructor must run for soundness reasons.

Maybe we can adjust to something like?

> Note that while code can safely prevent a destructor from running, they are still guaranteed to automatically run in places defined by this document (such as the end of scopes). Authors may rely on destructors running at these locations for soundness, but must not otherwise assume destructors are run.

Reply:
Ping @nbdd0121, was wondering if you'd be willing to change the note to something like the suggestion given here?

Reply:
Sorry for not being responsive, I am quite busy recently.

I don't think it's good to explicit saying that "you can expect destructor at the end of scope to run for soundness", as I think it could create additional confusion, as these are really just part of control flow that users expect.

The purpose of this change is to keep its original intention while removing the possibility to interpret it as a permit for compilers to omit destructor calls.

Reply:
No worries, I was just checking in.

How about something like this?

> Note: Preventing a destructor from being run via `std::mem::forget` or other
> means is safe even if it has a type that isn't `'static`.
> Besides the places where destructors are guaranteed to run as defined by
> this document, types may **not** safely rely on a destructor being run for
> soundness.

I'm trying to blend the different ideas here:

* It should always be safe to call `forget`.
* There are places where destructors are guaranteed to run, and you may rely on that.
* Types may not rely on their destructor being run for soundness.

I'm agree with Mark about the "publicly exposed APIs", so I'm trying to avoid that phrasing.

Reply:
Sounds very good. I have applied your suggestion.